File: web/src/protocol/rpcFrames.ts
/// rpc-frame 分片传输 — TS 移植自 lib/protocol/rpc_frames.dart + fragment_assembler.dart。
/// 逻辑消息切块保证每个 JSON 信封 <512KB；每消息带 crc32 校验；
/// 对端用 rpc-frame-ack 确认。

import {
  RPC_FRAME_MAX_FRAGMENT_PAYLOAD_BYTES,
  RPC_FRAME_MAX_MESSAGE_BYTES,
  RPC_FRAME_MAX_FRAGMENTS,
} from './constants'
import { crc32Hex } from './crc32'

export interface RpcFramePayload {
  [k: string]: unknown
}

interface Assembly {
  parts: (Uint8Array | null)[]
  received: number
  fragmentCount: number
  messageBytes: number
  checksum: string | null
  createdAt: number
}

export class RpcFrames {
  private seq = 0
  private messageSeq = 0
  private assemblies = new Map<number, Assembly>()
  private cleanupTimer: ReturnType<typeof setInterval>

  constructor(
    private opts: {
      bridgeSessionId: string
      bridgeGeneration?: number
      recoveryId?: string
      send: (payload: RpcFramePayload) => void
      onMessage: (bytes: Uint8Array) => void
      onLog?: (line: string) => void
    },
  ) {
    this.cleanupTimer = setInterval(() => this.purge(), 30_000)
  }

  private get identity(): RpcFramePayload {
    return {
      bridgeSessionId: this.opts.bridgeSessionId,
      ...(this.opts.bridgeGeneration != null
        ? { bridgeGeneration: this.opts.bridgeGeneration }
        : {}),
      ...(this.opts.recoveryId ? { recoveryId: this.opts.recoveryId } : {}),
    }
  }

  sendMessage(bytes: Uint8Array): void {
    if (bytes.length === 0) throw new Error('empty rpc message')
    if (bytes.length > RPC_FRAME_MAX_MESSAGE_BYTES) throw new Error('rpc message too large')
    const messageSeq = ++this.messageSeq
    const checksum = crc32Hex(bytes)
    const fragmentCount = Math.ceil(bytes.length / RPC_FRAME_MAX_FRAGMENT_PAYLOAD_BYTES)
    if (fragmentCount > RPC_FRAME_MAX_FRAGMENTS) throw new Error('fragment limit exceeded')
    for (let i = 0; i < fragmentCount; i++) {
      const start = i * RPC_FRAME_MAX_FRAGMENT_PAYLOAD_BYTES
      const end = Math.min(start + RPC_FRAME_MAX_FRAGMENT_PAYLOAD_BYTES, bytes.length)
      const chunk = bytes.subarray(start, end)
      this.seq += 1
      this.opts.send({
        zcode_type: 'rpc-frame',
        ...this.identity,
        seq: this.seq,
        messageSeq,
        fragmentIndex: i,
        fragmentCount,
        messageBytes: bytes.length,
        checksum: { algorithm: 'crc32', value: checksum },
        dataBase64: uint8ToBase64(chunk),
      })
    }
  }

  /** 喂入 relay payload（只认 rpc-frame(-ack)；其余忽略）。 */
  accept(payload: RpcFramePayload): void {
    const type = payload['zcode_type']
    if (type !== 'rpc-frame' && type !== 'rpc-frame-ack') return
    if (payload['bridgeSessionId'] !== this.opts.bridgeSessionId) return

    if (type === 'rpc-frame-ack') return // 本端不发大消息时 ack 无需处理

    const messageSeq = payload['messageSeq'] as number | undefined
    const fragmentIndex = payload['fragmentIndex'] as number | undefined
    const fragmentCount = payload['fragmentCount'] as number | undefined
    const messageBytes = payload['messageBytes'] as number | undefined
    const dataBase64 = payload['dataBase64'] as string | undefined
    const checksum = (payload['checksum'] as Record<string, unknown> | undefined)?.value as
      | string
      | undefined
    if (
      messageSeq == null ||
      fragmentIndex == null ||
      fragmentCount == null ||
      messageBytes == null ||
      dataBase64 == null
    ) {
      return
    }
    if (
      messageSeq < 0 ||
      fragmentCount < 1 ||
      fragmentCount > RPC_FRAME_MAX_FRAGMENTS ||
      fragmentIndex < 0 ||
      fragmentIndex >= fragmentCount ||
      messageBytes < 1 ||
      messageBytes > RPC_FRAME_MAX_MESSAGE_BYTES
    ) {
      return
    }

    let chunk: Uint8Array
    try {
      chunk = base64ToUint8(dataBase64)
    } catch {
      return
    }
    if (chunk.length > RPC_FRAME_MAX_FRAGMENT_PAYLOAD_BYTES) return

    const existing = this.assemblies.get(messageSeq)
    if (
      existing &&
      (existing.fragmentCount !== fragmentCount ||
        existing.messageBytes !== messageBytes ||
        existing.checksum !== checksum)
    ) {
      this.assemblies.delete(messageSeq)
      return
    }
    let assembly = existing
    if (!assembly) {
      assembly = {
        parts: new Array(fragmentCount).fill(null),
        received: 0,
        fragmentCount,
        messageBytes,
        checksum: checksum ?? null,
        createdAt: Date.now(),
      }
      this.assemblies.set(messageSeq, assembly)
    }
    if (fragmentIndex >= 0 && fragmentIndex < assembly.fragmentCount) {
      if (assembly.parts[fragmentIndex] === null) assembly.received += 1
      assembly.parts[fragmentIndex] = chunk
    }
    if (assembly.received === assembly.fragmentCount) {
      this.assemblies.delete(messageSeq)
      const message = concat(assembly.parts.filter((p): p is Uint8Array => p !== null))
      if (message.length !== assembly.messageBytes) {
        this.opts.onLog?.(`[rpc] message ${messageSeq} size mismatch`)
      } else if (assembly.checksum === null || crc32Hex(message) === assembly.checksum) {
        this.opts.send({
          zcode_type: 'rpc-frame-ack',
          ...this.identity,
          ackMessageSeq: messageSeq,
        })
        this.opts.onMessage(message)
      } else {
        this.opts.onLog?.(`[rpc] message ${messageSeq} checksum mismatch`)
      }
    }
  }

  private purge(): void {
    const now = Date.now()
    for (const [seq, a] of this.assemblies) {
      if (now - a.createdAt > 60_000) {
        this.assemblies.delete(seq)
        this.opts.onLog?.(`[rpc] purged stale assembly ${seq}`)
      }
    }
  }

  dispose(): void {
    clearInterval(this.cleanupTimer)
    this.assemblies.clear()
  }
}

function concat(parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((n, p) => n + p.length, 0)
  const out = new Uint8Array(total)
  let pos = 0
  for (const p of parts) {
    out.set(p, pos)
    pos += p.length
  }
  return out
}

export function uint8ToBase64(bytes: Uint8Array): string {
  let bin = ''
  const CHUNK = 0x8000 // 32K 分块转换，避免 apply 栈溢出
  for (let i = 0; i < bytes.length; i += CHUNK) {
    bin += String.fromCharCode(...bytes.subarray(i, i + CHUNK))
  }
  return btoa(bin)
}

export function base64ToUint8(b64: string): Uint8Array {
  const bin = atob(b64)
  const out = new Uint8Array(bin.length)
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i)
  return out
}

File: lib/protocol/rpc_frames.dart
import 'dart:async';
import 'dart:convert';
import 'dart:typed_data';

import 'constants.dart';
import 'crc32.dart';
import 'fragment_assembler.dart';

/// rpc-frame fragmentation transport over relay payloads.
///
/// Logical messages are split so each JSON envelope stays below 1 MiB; each
/// message carries a crc32 checksum; the peer acknowledges assembled
/// messages with `rpc-frame-ack`.
class RpcFrames {
  final String bridgeSessionId;
  final int? bridgeGeneration;
  final String? recoveryId;
  final void Function(Map<String, dynamic> payload) send;
  final void Function(Uint8List message) onMessage;
  final void Function(String line)? onLog;

  int _seq = 0;
  int _messageSeq = 0;

  final _assemblies = <int, _Assembly>{};
  Timer? _cleanup;

  RpcFrames({
    required this.bridgeSessionId,
    required this.send,
    required this.onMessage,
    this.bridgeGeneration,
    this.recoveryId,
    this.onLog,
  }) {
    _cleanup = Timer.periodic(const Duration(seconds: 30), (_) => _purge());
  }

  Map<String, dynamic> get _identity => {
        'bridgeSessionId': bridgeSessionId,
        if (bridgeGeneration != null) 'bridgeGeneration': bridgeGeneration,
        if (recoveryId != null) 'recoveryId': recoveryId,
      };

  void sendMessage(Uint8List bytes) {
    if (bytes.isEmpty) throw StateError('empty rpc message');
    if (bytes.length > rpcFrameMaxMessageBytes) {
      throw StateError('rpc message too large');
    }
    final messageSeq = ++_messageSeq;
    final checksum = Crc32.hexOf(bytes);
    final fragmentCount = (bytes.length + rpcFrameMaxFragmentPayloadBytes - 1) ~/
        rpcFrameMaxFragmentPayloadBytes;
    if (fragmentCount > rpcFrameMaxFragments) {
      throw StateError('fragment limit exceeded');
    }
    for (var i = 0; i < fragmentCount; i++) {
      final start = i * rpcFrameMaxFragmentPayloadBytes;
      final end = (start + rpcFrameMaxFragmentPayloadBytes) > bytes.length
          ? bytes.length
          : start + rpcFrameMaxFragmentPayloadBytes;
      final chunk = Uint8List.sublistView(bytes, start, end);
      _seq += 1;
      send({
        'zcode_type': 'rpc-frame',
        ..._identity,
        'seq': _seq,
        'messageSeq': messageSeq,
        'fragmentIndex': i,
        'fragmentCount': fragmentCount,
        'messageBytes': bytes.length,
        'checksum': {'algorithm': 'crc32', 'value': checksum},
        'dataBase64': base64.encode(chunk),
      });
    }
  }

  /// Feed a relay payload (rpc-frame(-ack) only; others are ignored).
  void accept(Map<String, dynamic> payload) {
    final type = payload['zcode_type'];
    if (type != 'rpc-frame' && type != 'rpc-frame-ack') return;
    if (payload['bridgeSessionId'] != bridgeSessionId) return;

    final messageSeq = (payload['messageSeq'] as num?)?.toInt();
    final fragmentIndex = (payload['fragmentIndex'] as num?)?.toInt();
    final fragmentCount = (payload['fragmentCount'] as num?)?.toInt();
    final messageBytes = (payload['messageBytes'] as num?)?.toInt();
    final dataBase64 = payload['dataBase64'] as String?;
    final checksum = (payload['checksum'] as Map?)?['value'] as String?;
    if (messageSeq == null ||
        fragmentIndex == null ||
        fragmentCount == null ||
        messageBytes == null ||
        dataBase64 == null) {
      return;
    }
    if (messageSeq < 0 ||
        fragmentCount < 1 ||
        fragmentCount > rpcFrameMaxFragments ||
        fragmentIndex < 0 ||
        fragmentIndex >= fragmentCount ||
        messageBytes < 1 ||
        messageBytes > rpcFrameMaxMessageBytes) {
      return;
    }

    Uint8List chunk;
    try {
      chunk = base64.decode(dataBase64);
    } catch (_) {
      return;
    }
    if (chunk.length > rpcFrameMaxFragmentPayloadBytes) return;

    final existing = _assemblies[messageSeq];
    if (existing != null &&
        (existing.fragmentCount != fragmentCount ||
            existing.messageBytes != messageBytes ||
            existing.checksum != checksum)) {
      _assemblies.remove(messageSeq);
      return;
    }
    final assembly = _assemblies.putIfAbsent(
      messageSeq,
      () => _Assembly(fragmentCount, messageBytes, checksum),
    );
    assembly.add(fragmentIndex, chunk);
    if (assembly.isComplete) {
      _assemblies.remove(messageSeq);
      final message = assembly.assemble();
      if (message.length != assembly.messageBytes) {
        onLog?.call('[rpc] message $messageSeq size mismatch');
      } else if (checksum == null || Crc32.hexOf(message) == checksum) {
        send({
          'zcode_type': 'rpc-frame-ack',
          ..._identity,
          'ackMessageSeq': messageSeq,
        });
        onMessage(message);
      } else {
        onLog?.call('[rpc] message $messageSeq checksum mismatch');
      }
    }
    return;
  }

  void _purge() {
    final now = DateTime.now();
    final stale = <int>[];
    _assemblies.forEach((seq, a) {
      if (now.difference(a.createdAt).inSeconds > 60) stale.add(seq);
    });
    for (final seq in stale) {
      _assemblies.remove(seq);
      onLog?.call('[rpc] purged stale assembly $seq');
    }
  }

  void dispose() {
    _cleanup?.cancel();
    _assemblies.clear();
  }
}

/// rpc 重组元数据 + 共享分片核心（缓冲/计数/拼装都在 FragmentAssembler）。
class _Assembly {
  final FragmentAssembler core;
  final int messageBytes;
  final String? checksum;

  _Assembly(int fragmentCount, this.messageBytes, this.checksum)
      : core = FragmentAssembler(fragmentCount);

  int get fragmentCount => core.fragmentCount;
  DateTime get createdAt => core.createdAt;
  void add(int index, Uint8List data) => core.add(index, data);
  bool get isComplete => core.isComplete;
  Uint8List assemble() => core.assemble();
}
